nirvana.go

```diff
@@ -17,54 +17,247 @@ limitations under the License.
 package nirvana
 
 import (
-	"runtime"
+	"fmt"
+	"net/http"
 
+	"github.com/caicloud/nirvana/definition"
+	"github.com/caicloud/nirvana/errors"
 	"github.com/caicloud/nirvana/log"
 	"github.com/caicloud/nirvana/service"
-	"github.com/caicloud/nirvana/utils/metrics"
-	"github.com/caicloud/nirvana/utils/profiling"
 )
 
-// Config describes configuration of service.Server
+// Server is a complete API server.
+// The server contains a router to handle all requests form clients.
+type Server interface {
+	// Serve starts to listen and serve requests.
+	// The method won't return except an error occurs.
+	Serve() error
+}
+
+// Config describes configuration of server.
 type Config struct {
-	EnableProfiling bool
-	// Requires generic profiling enabled
-	EnableContentionProfiling bool
-	EnableMetrics             bool
-	// Path of prometheus metrics endpoint, default to '/metrics'
-	MetricsPath     string
-	Logger          log.Logger
-	ConfigureServer func(service.Server)
-}
-
-// New creates a new service.Server
-func New(c *Config) (service.Server, error) {
-	if err := service.RegisterDefaultEnvironment(); err != nil {
-		panic(err)
-	}
-	var s service.Server
-	if c.ConfigureServer == nil {
-		s = service.NewDefaultServer()
+	// IP is the ip to listen. Empty means `0.0.0.0`.
+	IP string
+	// Port is the port to listen.
+	Port uint16
+	// Logger is used to output info inside framework.
+	Logger log.Logger
+	// Descriptors contains all APIs.
+	Descriptors []definition.Descriptor
+	// Filters is http filters.
+	Filters []service.Filter
+	// Modifiers is definition modifiers
+	Modifiers service.DefinitionModifiers
+	// configSet contains all configurations of plugins.
+	configSet map[string]interface{}
```

Comment:
Why not just `plugins` if this is all about configurations of plugins?  configSet is too generic

Reply:
`configSet` only stores configs of plugins. If it's named `plugins`, may cause misunderstanding.